### PR TITLE
commit template/templater/formatter: Label obsolete commits

### PR DESCRIPTION
**Pull request by ilyagr**
This PR has two separate features:

- Support for a "normal" color in the formatter that can override the color of a parent's style.
- Obsolete commits are now labeled as such. This can be useful when, e.g., doing `jj log commit_id`.

If needed, I can split this into two PRs (this may require changing the styling of obsolete commits).

Here is an example of test of `jj obslog` (note: the underlines were later removed):

image

`jj log`:

image

<details><summary>Previous version for comparison</summary>

image

</details> 

I also considered the wording `(Was change_id)` or `change_id (old)`, but I settled on `change_id (obs)` + color.

# Checklist

If applicable:
- [x] I have updated `CHANGELOG.md`
- [x] I have updated the documentation (README.md, docs/, demos/)
- [x] I have updated the config schema (src/config-schema.json)
- [x] I have added tests to cover my changes

### CODE REVIEW

File: src/config/colors.toml [hunk 1]
```diff
@@ -42,6 +44,7 @@
 "working_copy git_refs" = "bright magenta"
 "working_copy divergent" =  "bright red"
 "working_copy divergent change_id"="bright red"
+"working_copy obsolete change_id"="bright normal"  # Not currently achievable
```
Comment (ilyagr) [inline src/config/colors.toml:47]: Note to self: replace this with normal. Bright normal does not exist

File: tests/test_obslog_command.rs [hunk 2]
```diff
@@ -33,14 +33,27 @@ fn test_obslog_with_or_without_diff() {
     insta::assert_snapshot!(stdout, @r###"
     @  rlvkpnrzqnoo test.user@example.com 2001-02-03 04:05:10.000 +07:00 66b42ad36073
     │  my description
-    ◉  rlvkpnrzqnoo test.user@example.com 2001-02-03 04:05:09.000 +07:00 af536e5af67e conflict
+    ◉  rlvkpnrzqnoo (obs) test.user@example.com 2001-02-03 04:05:09.000 +07:00 af536e5af67e conflict
```
Comment (yuja) [inline tests/test_obslog_command.rs:36]: `(obs)` is unclear to me, but I feel `(obsolete)` would be too verbose (especially in obslog), so I have no better idea.

Just fyi, in Mercurial, obsolete commit is represented as `x` in graph.

Comment (martinvonz) [reply tests/test_obslog_command.rs:36]: I feel like "hidden" is better than "obsolete". That would match `visible_heads()` (which I recently renamed `heads()` to).

And yes, it might be a good idea to indicate them in the graph. Even if we choose to present it only there by default, it's probably good to also have the information in the template so users can access it even when using `--no-graph`.

Comment (necauqua) [reply tests/test_obslog_command.rs:36]: Yet again bringing up git-branchless, such commits are marked with × in place of ◉ in it

File: src/formatter.rs [hunk 3]
```diff
@@ -371,6 +371,7 @@ fn rules_from_config(config: &config::Config) -> Result<Rules, config::ConfigErr
 
 fn color_for_name(color_name: &str) -> Option<Color> {
     match color_name {
+        "normal" => Some(Color::Reset),
```
[the remark this answers is not in the document]
Comment (ilyagr) [reply src/formatter.rs:374]: Agreed, "default" is better. Thanks!

File: src/commit_templater.rs [hunk 4]
```diff
@@ -248,6 +248,10 @@ fn build_commit_keyword_opt<'repo>(
             let maybe_entries = repo.resolve_change_id(commit.change_id());
             maybe_entries.map_or(0, |entries| entries.len()) > 1
         })),
+        "obsolete" => language.wrap_boolean(wrap_fn(property, |commit| {
+            let maybe_entries = repo.resolve_change_id(commit.change_id());
+            maybe_entries.map_or(true, |entries| !entries.iter().any(|id| id == commit.id()))
```
[the remark this answers is not in the document]
Comment (ilyagr) [reply src/commit_templater.rs:253]: Thanks!

File: tests/test_obslog_command.rs (hunk 2 above)
Comment (ilyagr) [reply tests/test_obslog_command.rs:36]: > (obs) is unclear to me, but I feel (obsolete) would be too verbose

I think I'll go with "hidden" for now, as Martin suggested, since it's short. I don't like that word, but Sapling and git-branchless use it and I don't have a better idea at the moment.

> Yet again bringing up git-branchless, such commits are marked with × in place of ◉ in it

Agreed, this would be great. It doesn't seem like a trivial feature to add at the moment AFAIK because of how the graph is implemented, but it could certainly be done.

File: src/config/templates.toml [hunk 5]
```diff
@@ -12,7 +12,9 @@ label(if(current_working_copy, "working_copy"),
     separate(" ",
       if(divergent,
         label("divergent", format_short_change_id(change_id) ++ "??"),
```
Comment (ilyagr) [inline src/config/templates.toml:14]: I actually realized there is a subtle minor bug here: a commit could have a divergent change id and yet be obsolete.

It's easy to fix it here. I'm not sure if the definition of `divergent` condition should be changed, I'll probably just add a comment there for now.

I'll also see if it's not too hard to write a test. Since the bug is so minor, I might leave it for another PR.

Comment (ilyagr) [reply src/config/templates.toml:14]: Actually, I think it's simpler (in terms of fixing and testing) to treat this as a bug in the "divergent" condition as opposed to this template. This also makes sense as "divergent" is a property of a commit here (not of a change id).

**Comment by ilyagr** [comment]
By the way, thank you @yuja, working with these in `commit_templater.rs` is so much easier now than it used to be!

File: src/config/templates.toml (hunk 5 above)
Comment (ilyagr) [reply src/config/templates.toml:14]: I'm now thinking these should be written in white as

```
xyzzy?? hidden (normal colored email, date, etc)
```

since it makes sense to print divergent change ids differently.

Comment (ilyagr) [reply src/config/templates.toml]: Resolved in a new commit.

**Comment by ilyagr** [comment]
In case you didn't see my discussion with myself, I added a new commit to show hidden commits with divergent change ids properly.

**Comment by ilyagr** [comment]
I made one more change: I think it looks better without underlines (as long as the tails of the prefixes are grey).

image

File: src/commit_templater.rs [hunk 6]
```diff
@@ -248,6 +248,10 @@ fn build_commit_keyword_opt<'repo>(
             let maybe_entries = repo.resolve_change_id(commit.change_id());
             maybe_entries.map_or(0, |entries| entries.len()) > 1
         })),
+        "obsolete" => language.wrap_boolean(wrap_fn(property, |commit| {
```
Comment (yuja) [inline src/commit_templater.rs:251]: The keyword should also be renamed to "hidden" (or "visible" if we prefer a positive term)?

I think "hidden" is technically correct than "obsolete".

File: CHANGELOG.md [hunk 7]
```diff
@@ -73,7 +73,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 * The progress display on `jj git clone/fetch` now includes the downloaded size.
 
 * The formatter now supports a "default" color that can override another color
-  defined by a parent style.
```
Comment (yuja) [inline CHANGELOG.md]: nit: unrelated change or deleted by mistake?

File: src/config/templates.toml [hunk 8]
```diff
@@ -10,10 +10,13 @@ log = '''
...
-        label("divergent", format_short_change_id(change_id) ++ "??"),
-        if(obsolete,
-          label("obsolete", format_short_change_id(change_id) ++ " hidden"),
+      if(obsolete,
+        label("obsolete",
+          (format_short_change_id(change_id) ++
+            if(divergent, label("divergent", "??"), "") ++
+            " hidden")),
+        if(divergent,
+          label("divergent", format_short_change_id(change_id) ++ "??"),
```
Comment (yuja) [inline src/config/templates.toml:19]: I don't follow closely, but maybe you can build labels and contents separately:
```
label(
  separate(" ", if(divergent, "divergent"), if(obsolete, "obsolete")),
  format_short_change_id(change_id) ++ if(divergent, "??") ...
```

File: src/commit_templater.rs (hunk 6 above)
Comment (ilyagr) [reply src/commit_templater.rs:251]: I was using "obsolete" with the thought that, eventually, these can be garbage collected.

In my mind, a hidden commit is one that's not shown by my default revset (so, most commits on the main branch).

But it seems people don't share this sentiment. If I don't hear anything, I will change it to "visible" or "hidden".

File: CHANGELOG.md (hunk 7 above)
Comment (ilyagr) [reply CHANGELOG.md]: Good catch, thanks!

File: src/config/templates.toml (hunk 8 above)
Comment (ilyagr) [reply src/config/templates.toml:19]: I didn't realize you can specify more than one label at a time. I'll try this, will need to update the colors accordingly.

File: src/commit_templater.rs (hunk 6 above)
Comment (yuja) [reply src/commit_templater.rs:251]: Maybe I'm too used to using Mercurial, but I think "obsolete" is a commit superseded by something else (or abandoned), which could be temporarily un-"hidden" by visible descendants (causing divergence.)

> a hidden commit is one that's not shown by my default revset

That's an interesting point. Maybe "hidden" is too weak to describe a permanent property? Show/hide is a common word in tree-like UI to temporarily switch the visibility.

Comment (yuja) [reply src/commit_templater.rs:251]: > That's exactly what these are, as far as I can tell.

Well, I meant anything other than the latest entry in obslog are "obsolete", but not always "hidden". Perhaps, my "hidden" definition is identical to your "obsolete".

File: tests/test_obslog_command.rs (hunk 2 above)
Comment (martinvonz) [reply tests/test_obslog_command.rs]: > Agreed, this would be great. It doesn't seem like a trivial feature to add at the moment AFAIK because of how the graph is implemented, but it could certainly be done.

By the way, I'm soon going to work on defining a revset for immutable commits, and another revset for getting shorter commit/change id prefixes. That work might make it easier to also check if a given commit is visible.

File: src/commit_templater.rs (hunk 6 above)
Comment (martinvonz) [reply src/commit_templater.rs:251]: I think we should go with "hidden" for now, mostly because it's consistent with `visible_heads()` (and `unobsolete_heads()` or `valid_heads()` doesn't sounds great either).

Comment (ilyagr) [reply src/commit_templater.rs:251]: OK, let's go with hidden.

File: src/config/templates.toml (hunk 8 above)
Comment (ilyagr) [reply src/config/templates.toml]: Yes, I think this works better. Now colors.toml is more complex, but that's fine.

**Comment by ilyagr** [comment]
Here's a summary of the latest changes:

The docs for "divergent" were good already, but I added a word.

I changed the commit template, which made `colors.toml` more complicated.

I'm also wondering about whether to make "hidden" bold and reorganize `colors.toml` a bit. I might make that a separate PR, since there are annoying seemingly no-op changes to the color output. See the last commit in https://github.com/martinvonz/jj/compare/main...ilyagr:jj:obs2.

File: src/formatter.rs [hunk 9]
```diff
@@ -861,6 +862,36 @@ mod tests {
         @"invalid color: bloo");
     }
 
+    #[test]
+    fn test_color_formatter_normal_color() {
+        // Normal color resets the color. It can reset only background or only
+        // foreground.
```
Comment (yuja) [inline src/formatter.rs:868]: nit: s/Normal/Default/

**Review by yuja — APPROVED** [review]
LGTM, thanks.

**Comment by ilyagr** [comment]
Thanks! This is stuck on the cargo deny check, but that's hopefully temporary.

File: tests/test_obslog_command.rs (hunk 2 above)
Comment (ilyagr) [reply tests/test_obslog_command.rs]: Yeah, it would make a lot of sense to use commit markers to distinguish between immutable, mutable, and obsolete commits. They could be, for example, `◆, ◉, ×`. Another nice symbol is `■`. I'm mostly taking these from https://www.w3.org/TR/xml-entity-names/025.html
